test/apps/iodemo/io_demo.cc

```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
         XFER_TYPE_RECV
     } xfer_type_t;
 
+    class BufferChunk {
+    public:
+        BufferChunk(size_t size, MemoryPool<BufferChunk> *pool) :
+            _capacity(size), _buffer(memalign(ALIGNMENT, size)), _size(0),
+            _pool(*pool) {
+            if (_buffer == NULL) {
+                throw std::bad_alloc();
+            }
+        }
+
+        void release() {
+            _pool.put(this);
+        }
+
+        inline void *buffer(size_t offset = 0) const {
+            return (uint8_t*)_buffer + offset;
+        }
+
+        inline void resize(size_t size) {
+            assert(size <= _capacity);
+            _size = size;
+        }
+
+        inline size_t size() const {
+            return _size;
+        }
+
+        ~BufferChunk() {
+            free(_buffer);
+        }
+
+    public:
+        const size_t              _capacity;
+
+    private:
+        void *                    _buffer;
+        size_t                    _size;
+        MemoryPool<BufferChunk> & _pool;
+    };
+
+    class BufferIov {
+    public:
+        BufferIov(size_t size, MemoryPool<BufferIov> *pool) : _pool(*pool) {
+            _iov.reserve(size);
+        }
+
+        size_t size() const {
+            return _iov.size();
+        }
+
+        void init(size_t data_size, MemoryPool<BufferChunk> &chunk_pool,
+                  uint32_t sn, bool fill) {
+            assert(_iov.empty());
+
+            BufferChunk *chunk = chunk_pool.get();
+            _iov.resize(get_chunk_cnt(data_size, chunk->_capacity));
+
+            size_t remaining = init_chunk(0, chunk, data_size);
+            for (size_t i = 1; i < _iov.size(); ++i) {
+                remaining = init_chunk(i, chunk_pool.get(), remaining);
+            }
+
+            assert(remaining == 0);
+
+            if (fill) {
+                fill_data(sn);
+            }
+        }
+
+        inline BufferChunk& operator[](size_t i) const {
+            return *_iov[i];
+        }
+
+        void release() {
+            while (!_iov.empty()) {
+                _iov.back()->release();
+                _iov.pop_back();
+            }
+
+            _pool.put(this);
+        }
+
+        uint16_t chk_sum() const {
+            return chk_hdr()->chk_sum;
+        };
+
+        uint32_t sn() const {
+            return tr_hdr()->sn;
+        }
+
+        inline uint16_t calc_chk_sum() const {
+            uint16_t chk_sum = ucs_crc16(tr_hdr(), (uintptr_t)tr_hdr() -
+                                                   (uintptr_t)_iov[0]->buffer());
+
+            for (size_t i = 1; i < _iov.size(); ++i) {
+                chk_sum ^= ucs_crc16(_iov[i]->buffer(), _iov[i]->size());
```

Comment:
maybe add crc16 function to Buffer class

Reply:
Buffer does not know about possible header in first chunk, this is why it's handled in Iov

Reply:
i mean, each buffer will compute its check sum independently and then xor all in this func
BTW, why need special case for buffer 0, and not just init chk_sum to 0?

Reply:
crc will be stored in header of first buffer, buffers is not aware about this and result crc will be wrong
> BTW, why need special case for buffer 0, and not just init chk_sum to 0?

exactly by this reason, crc of first buffer must be calculated with header offset

Reply:
does it include sn?

Reply:
yes, tr_hdr() contains this, initialized by `fill_data` on sender side or received with data on receiver side